File: solutions/fsharp/protein-translation/2/ProteinTranslation.fs
module ProteinTranslation

open System

let codonToProtein = 
    function
    | "AUG" -> "Methionine"
    | "UGG" -> "Tryptophan"
    | "UGU" | "UGC" -> "Cysteine"
    | "UUA" | "UUG" -> "Leucine"
    | "UUC" | "UUU" -> "Phenylalanine"
    | "UAU" | "UAC" -> "Tyrosine"
    | "UAA" | "UAG" | "UGA" -> "STOP"
    | "UCU" | "UCC" | "UCA" | "UCG" -> "Serine"
    | _ -> failwith "Invalid codon"

let proteins (rna: string) =
    rna 
    |> Seq.chunkBySize 3 
    |> Seq.map (String >> codonToProtein)
    |> Seq.takeWhile (fun str -> str <> "STOP")
    |> Seq.toList

File: solutions/python/protein-translation/2/protein_translation.py
codons = {
    "AUG": "Methionine",
    "UUU": "Phenylalanine",
    "UUC": "Phenylalanine",
    "UUA": "Leucine",
    "UUG": "Leucine",
    "UCU": "Serine",
    "UCC": "Serine",
    "UCA": "Serine",
    "UCG": "Serine",
    "UAU": "Tyrosine",
    "UAC": "Tyrosine",
    "UGU": "Cysteine",
    "UGC": "Cysteine",
    "UGG": "Tryptophan",
    "UAA": "STOP",
    "UAG": "STOP",
    "UGA": "STOP",
}


def proteins(strand):
    protein_strand = [
        codons[strand[index: index + 3]]
        for index in range(0, len(strand), 3)
    ]
    return protein_strand[
       : None if "STOP" not in protein_strand else protein_strand.index("STOP")
    ]
